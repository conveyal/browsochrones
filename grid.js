/** 
 * takes spatial data from GeoJSON and coerces it to regular grid
 * @author mattwigway
 */

import extent from 'geojson-extent'
import inside from 'turf-inside'
import {pixelToLat, pixelToLon, lonToPixel, latToPixel} from './mercator'

const HEADER_SIZE = 24 // bytes, must be multiple of 8 as we're using a float64array (no clue why)

/**
 * Take geojson point or polygon data and convert it to a grid. Return a map from keys to raw array buffers formatted as follows
 * (4 byte int) zoom level
 * (4 byte int) west (x) offset
 * (4 byte int) north (y) offset
 * (4 byte int) width
 * (4 byte int) height
 * 4 bytes padding to make header a multiple of 8 bytes
 * repeated 8 byte _double_ values for each pixel, rows first, delta-coded
 */
export default function grid (data, zoom) {
  // for now assume that all numeric properties are to be included, and that all features have the same properties
  // geojson-extent has a fit if some features don't have properties
  data.features = data.features.filter(f => f.properties != null)
  let exemplar = data.features[0]

  console.log(exemplar.properties)

  // figure out bounding box
  let bbox = extent(data)
  // bbox is w, s, e, n
  let west = lonToPixel(bbox[0], zoom)
  let south = latToPixel(bbox[1], zoom)
  let east = lonToPixel(bbox[2], zoom)
  let north = latToPixel(bbox[3], zoom)
  let width = east - west
  let height = south - north // +y is south

  console.log(`n ${north} e ${east} s ${south} w ${west} width ${width} height ${height}`)

  let bufSizeBytes = HEADER_SIZE + 8 * width * height

  let out = new Map()

  for (var key in exemplar.properties) {
    if (!exemplar.properties.hasOwnProperty(key)) continue // shouldn't be an issue, as there should be no prototypes, but who knows?

    var val = exemplar.properties[key]

    if (isFinite(val)) {
      let buf = new ArrayBuffer(bufSizeBytes)
      // fill the first few value
      let dv = new DataView(buf)
      dv.setInt32(0, zoom, true)
      dv.setInt32(4, west, true)
      dv.setInt32(8, north, true)
      dv.setInt32(12, width, true)
      dv.setInt32(16, height, true)

      let arr = new Float64Array(buf, HEADER_SIZE)

      out.set(key, arr)
    }
  }

  // loop over all features, accumulate to grid
  data.features.forEach(feat => {
    if (feat.properties === undefined || feat.geometry === undefined) return

    // figure out relevant pixels for this feature and how much of the feature they overlap
    let pixels = []
    if (feat.geometry.type === 'Polygon') {
      let fbbox = extent(feat.geometry)
      let fwest = lonToPixel(fbbox[0], zoom)
      let fsouth = latToPixel(fbbox[1], zoom)
      let feast = lonToPixel(fbbox[2], zoom)
      let fnorth = latToPixel(fbbox[3], zoom)

      for (let x = fwest; x <= feast; x++) {
        for (let y = fnorth; y <= fsouth; y++) {
          let pt = {
            geometry: {
              type: 'Point',
              coordinates: [pixelToLon(x, zoom), pixelToLat(y, zoom)]
            }
          }

          if (inside(pt, feat)) {
            pixels.push((y - north) * width + (x - west))
          }
        }
      }

      if (pixels.length === 0) {
        pixels.push((fnorth - north) * width + fwest - west)
      }
    } else if (feat.geometry.type === 'Point') {
      let x = lonToPixel(feat.geometry.coordinates[0], zoom) - west
      let y = latToPixel(feat.geometry.coordinates[1], zoom) - north
      pixels.push(y * width + x)
    } else {
      console.log('Attempt to calculate accessibility to unsupported feature type ' + feat.geometry.type)
      return
    }

    // todo pycnoplactic mapping
    let weight = 1 / pixels.length

    out.forEach((array, key) => {
      // TODO once we have a weight-per-pixel this won't work
      let val = feat.properties[key] * weight

      if (isNaN(val)) return

      pixels.forEach(p => array[p] += val)
    })
  })

  // delta-code values, extract raw array buffers
  let ret = new Map()
  out.forEach((array, key) => {
    // delta-code for efficient compression
    for (let i = 0, prev = 0; i < array.length; i++) {
      let current = array[i]
      array[i] = current - prev
      prev = current
    }

    // extract raw array buffer
    ret.set(key, array.buffer)
  })

  return ret
}
