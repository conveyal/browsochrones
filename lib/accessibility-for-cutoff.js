/**
 * Get the cumulative accessibility number for a cutoff from a travel time surface.
 * This function always calculates _average_ accessibility. Calculating best or worst case accessibility is computationally
 * complex because you must individually calculate accessibility for every minute, save all of those values, and then take a minumum.
 * (Saving the worst-case travel time to each pixel allows you to calculate a bound, but does not allow calculation of the true minimum, because
 * it is possible that all the worst-case travel times cannot appear simultaneously. Of course this comes back to the definition of your measure, and
 * how fungible you consider opportunities to be. )
 *
 * @param {Object} surface
 * @param {Object} grid
 * @returns {Number} accessibility
 */

export default function accessibilityForCutoff ({surface, grid}) {
  let query = surface.query
  let accessibility = 0

  for (let y = 0, pixel = 0; y < query.height; y++) {
    for (let x = 0; x < query.width; x++, pixel++) {
      let count = surface.access[pixel]

      // ignore unreached locations
      if (count === 0) continue

      let gridx = x + query.west - grid.west
      let gridy = y + query.north - grid.north

      // if condition below fails we're off the grid, value is zero, don't bother with calculations
      if (gridx >= 0 && gridx < grid.width && gridy >= 0 && gridy < grid.height) {
        // get value of this pixel from grid
        let val = grid.data[gridy * grid.width + gridx]
        let weight = count / surface.nMinutes
        accessibility += val * weight
      }
    }
  }

  return accessibility
}
