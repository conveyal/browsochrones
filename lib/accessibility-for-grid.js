/**
 * Get the cumulative accessibility number for a cutoff from a travel time surface.
 * This function always calculates _average_ accessibility. Calculating best or worst case accessibility is computationally
 * complex because you must individually calculate accessibility for every minute, save all of those values, and then take a minumum.
 * (Saving the worst-case travel time to each pixel allows you to calculate a bound, but does not allow calculation of the true minimum, because
 * it is possible that all the worst-case travel times cannot appear simultaneously. Of course this comes back to the definition of your measure, and
 * how fungible you consider opportunities to be.)
 *
 * The cutoff used is the cutoff that was specified in the surface generation. If you want a different cutoff you must regenerate the surface. The reason
 * for this is that we need to know at every minute whether each destination was reached within a certain amount of time. Storing this for every possible
 * cutoff is not feasible (the data become too large), so we only store it for a single cutoff during surface generation. However, calculating accessibility
 * for additional grids should only take milliseconds.
 *
 * @param {Object} surface
 * @param {Object} grid
 * @returns {Number} accessibility
 */

export default function accessibilityForGrid ({cutoff = 60, surface, grid}) {
  let query = surface.query
  let accessibility = 0

  // round cutoff to nearest 5 minutes, which is what we've calculated
  // index 0 is five minutes
  let cutoffIdx = Math.round(cutoff / 5) - 1

  for (let y = 0, pixel = 0; y < query.height; y++) {
    for (let x = 0; x < query.width; x++, pixel++) {
      let count = surface.access[cutoffIdx][pixel]

      // ignore unreached locations
      if (count === 0) continue

      let gridx = x + query.west - grid.west
      let gridy = y + query.north - grid.north

      // if condition below fails we're off the grid, value is zero, don't bother with calculations
      if (gridx >= 0 && gridx < grid.width && gridy >= 0 && gridy < grid.height) {
        // get value of this pixel from grid
        let val = grid.data[gridy * grid.width + gridx]
        let weight = count / surface.nMinutes
        accessibility += val * weight
      }
    }
  }

  return accessibility
}
