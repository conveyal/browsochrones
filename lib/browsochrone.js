
import accessibilityForCutoff from './accessibility-for-cutoff'
import isochroneTile from './isochrone-tile'
import getSurface from './get-surface'

const imageHeight = 256
const imageWidth = 256

export default class Browsochrone {
  drawTile (canvas, {x, y}, zoom) {
    const scaleFactor = Math.pow(2, zoom - this.query.zoom)

    // find top-left coords at zoom 10
    // NB hipsters would use bitshifts but bitwise operators in Javascript only work on 32-bit ints. Javascript does not have 64-bit integer types.
    const xoffset = Math.round(x * imageWidth / scaleFactor - this.query.west)
    const yoffset = Math.round(y * imageHeight / scaleFactor - this.query.north)
    // NB x and y offsets are now relative to query

    const ctx = canvas.getContext('2d')
    const imageData = ctx.createImageData(imageHeight, imageWidth)

    isochroneTile(imageData, {
      cutoffMinutes: 60,
      height: this.query.height,
      imageHeight,
      imageWidth,
      scaleFactor,
      surface: this.surface.surface,
      width: this.query.width,
      x,
      xoffset,
      y,
      yoffset
    })

    ctx.putImageData(imageData, 0, 0)
  }

  getAccessibilityForCutoff (cutoff = 60) {
    return accessibilityForCutoff({
      cutoff,
      surface: this.surface,
      which: this.which || 'AVERAGE'
    })
  }

  generateSurface ({grid, origin, query, stopTrees, which} = {}) {
    this.surface = getSurface({
      grid: grid || this.grid,
      origin: origin || this.origin,
      query: query || this.query,
      stopTreeCache: stopTrees || this.stopTrees,
      which: which || this.which || 'AVERAGE'
    })
    return this.surface
  }

  setGrid (arrayBuffer) {
    const arr = new Float64Array(arrayBuffer, 24)

    for (let i = 0, prev = 0; i < arr.length; i++) {
      arr[i] = (prev += arr[i])
    }

    const dv = new DataView(arrayBuffer)
    this.grid = {
      // parse header
      zoom: dv.getInt32(0, true),
      west: dv.getInt32(4, true),
      north: dv.getInt32(8, true),
      width: dv.getInt32(12, true),
      height: dv.getInt32(16, true),
      data: arr
    }
  }

  setOrigin (arrayBuffer) {
    this.origin = new Int32Array(arrayBuffer)
  }

  setQuery (json) {
    this.query = json
  }

  setStopTrees (arrayBuffer) {
    this.stopTrees = new Int32Array(arrayBuffer)
  }
}
