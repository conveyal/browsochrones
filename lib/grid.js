/**
 * A grid file.
 */

export default class Grid {
  constructor (data) {
    const array = new Int32Array(data, 4 * 5)
    const header = new Int32Array(data)

    for (let i = 0, prev = 0; i < array.length; i++) {
      array[i] = (prev += array[i])
    }
      // parse header
    this.zoom = header[0]
    this.west = header[1]
    this.north = header[2]
    this.width = header[3]
    this.height = header[4]
    this.data = array
  }
}
