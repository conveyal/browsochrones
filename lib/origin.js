/**
 * represents the data for a single origin, including a per-transit-stop index.
 * @author mattwigway
 */
export default class Origin {
	constructor (array) {
    this.data = array

    // build index
    let offset = 0
    let radius = array[offset++]
    offset += Math.pow(radius * 2 + 1, 2) // skip nontransit data
    let nStops = array[offset++]
    let nMinutes = array[offset++]

    this.index = new Int32Array(nStops)

    for (let stop = 0; stop < nStops; stop++) {
      this.index[stop] = offset

      // skip the times and path info
      offset += nMinutes * 2

      // read paths
      let nPaths = array[offset++]

      for (let path = 0; path < nPaths; path++) {
        let pathLen = array[offset++]
        offset += pathLen * 3 // 3 ints for each segment: board, pattern, and alight
      }
    }
  }
}
