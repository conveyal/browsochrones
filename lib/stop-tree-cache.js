/**
 * Cache trees from pixels to transit stops, including an index for where a particular pixel is located.
 * @author mattwigway
 */

/**
 * Construct a stop tree cache from a typed array
 *
 * @param {ArrayBuffer} data
 * @param {Number} size
 */
export function create (data, size) {
  const stopTreeCache = {}

  stopTreeCache.data = data
  stopTreeCache.index = new Int32Array(size)

  // function that reads from array performing bounds checks
  const read = (idx) => {
    if (idx >= data.length) {
      throw new Error('attempted to read past end of stop tree cache!')
    } else {
      return data[idx]
    }
  }

  // build the index and de-delta-code
  let pixelsRead = 0
  for (let i = 0, pixel = 0, stopId = 0, time = 0; i < data.length; pixel++) {
    stopTreeCache.index[pixel] = i
    const nStops = read(i++)

    for (let stopIdx = 0; stopIdx < nStops; stopIdx++) {
      stopId += read(i)
      data[i++] = stopId
      time += read(i)
      data[i++] = time
    }

    pixelsRead++
  }

  if (pixelsRead !== size) {
    throw new Error(`Expected to read ${size} pixels but read only ${pixelsRead}`)
  }

  return stopTreeCache
}
