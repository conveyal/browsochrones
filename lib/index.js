import WebWorkerPromiseInterface from 'web-worker-promise-interface'

import {create as createGridFunc} from './grid'
import {latToPixel, lonToPixel} from './mercator'
import workerHandlers from './worker-handlers'

const imageHeight = 256
const imageWidth = 256

function assert (e, msg) {
  if (!e) throw new Error(msg)
}

export default class Browsochrones extends WebWorkerPromiseInterface {
  originLoaded = false
  query = false
  stopTreesLoaded = false
  surfaceLoaded = false

  constructor (opts) {
    super(workerHandlers)
    const {origin, query, stopTrees, transitiveNetwork} = opts || {}

    if (origin) this.setOrigin(origin.data, origin.point)
    if (query) this.setQuery(query)
    if (stopTrees) this.setStopTrees(stopTrees)
    if (transitiveNetwork) this.setTransitiveNetwork(transitiveNetwork)

    this.drawTile = this.drawTile.bind(this)
  }

  /**
   * transfer a grid (as a raw arraybuffer) into this instance's worker. Note that this will neuter the grid, make a copy
   * before calling this function if you expect to use it elsewhere
   */
  putGrid (key, grid) {
    return this.work({
      command: 'putGrid',
      message: { key, grid }
    })
  }

  createTile = (coords, done) => {
    const tile = document.createElement('canvas')
    tile.height = imageHeight
    tile.width = imageWidth
    this.drawTile(tile, coords, coords.z)
      .then(() => done(null, tile))
      .catch((err) => done(err, tile))
    return tile
  }

  drawTile = (canvas, {x, y}, zoom) => {
    const scaleFactor = Math.pow(2, zoom - this.query.zoom)

    // find top-left coords at zoom 10
    // NB hipsters would use bitshifts but bitwise operators in Javascript only work on 32-bit ints. Javascript does not have 64-bit integer types.
    const xoffset = Math.round(x * imageWidth / scaleFactor - this.query.west)
    const yoffset = Math.round(y * imageHeight / scaleFactor - this.query.north)
    // NB x and y offsets are now relative to query

    const ctx = canvas.getContext('2d')
    const imageData = ctx.createImageData(imageHeight, imageWidth)

    return this.work({
      command: 'drawTile',
      message: {
        imageData,
        scaleFactor,
        xoffset,
        yoffset
      }
    }).then((message) => {
      ctx.putImageData(message, 0, 0)
      return message
    })
  }

  async getAccessibilityForGrid (grid, cutoff = 60) {
    assert(this.isLoaded(), 'Accessibility cannot be computed before generating a surface.')

    return this.work({
      command: 'accessibilityForGrid',
      message: {
        grid, cutoff
      }
    })
  }

  async generateDestinationData ({ from, to }) {
    assert(this.isLoaded(), 'Transitive data cannot be generated if Browsochrones is not fully loaded.')

    return this.work({
      command: 'generateDestinationData',
      message: {
        from,
        to
      }
    })
  }

  getPaths (point) {
    return this.work({
      command: 'getPaths',
      message: {
        point
      }
    })
  }

  getPath (path) {
    return this.work({
      command: 'getPath',
      message: {
        path
      }
    })
  }

  async generateSurface (grid, which = 'MEDIAN') {
    assert(this.isLoaded(), 'Surface cannot be generated if Browsochrones is not fully loaded.')

    return this.work({
      command: 'generateSurface',
      message: {
        grid,
        which
      }
    }).then((message) => {
      this.surfaceLoaded = true
      return message
    })
  }

  latLonToOriginPoint ({lat, lon, lng}) {
    assert(this.query, 'Cannot convert lat/lon without query.')
    const {north, west, zoom} = this.query
    const x = lonToPixel(lon || lng, zoom)
    const y = latToPixel(lat, zoom)
    const ret = {
      // TODO should these be rounded instead of floored?
      x: (x - west) | 0,
      y: (y - north) | 0
    }
    assert(this.pointInQueryBounds(ret), 'Point out of query bounds!')
    return ret
  }

  /**
   * @param {Number} pixel.x
   * @param {Number} pixel.y
   * @param {Number} currentZoom
   * @returns {Object} point
   */
  pixelToOriginPoint ({x, y}, currentZoom) {
    assert(this.query, 'Cannot convert point without query.')

    const {north, west, zoom} = this.query
    const scale = Math.pow(2, zoom - currentZoom)

    x = x * scale - west | 0
    y = y * scale - north | 0

    return {x, y}
  }

  /**
   * @param {Number} point.x
   * @param {Number} point.y
   * @returns {Boolean}
   */
  pointInQueryBounds ({x, y}) {
    const {height, width} = this.query
    return x >= 0 && x <= width && y >= 0 && y <= height
  }

  /**
   * @param {ArrayBuffer} data
   * @param {Object} point
   * @param {Number} point.x
   * @param {Number} point.y
   */
  setOrigin (arrayBuffer, point) {
    return this.work({
      command: 'setOrigin',
      message: {
        arrayBuffer,
        point
      },
      transferrable: [arrayBuffer]
    }).then(() => {
      this.originLoaded = true
    })
  }

  setQuery (json) {
    this.query = json
    return this.work({
      command: 'setQuery',
      message: {
        query: json
      }
    })
  }

  /**
   * Generate the stop tree cache. Must set the query first.
   *
   * @param {ArrayBuffer} data
   */
  async setStopTrees (arrayBuffer) {
    assert(this.query, 'Query must be loaded before generating the stop tree cache.')

    return this.work({
      command: 'setStopTreeCache',
      message: {
        arrayBuffer
      },
      transferrable: [arrayBuffer]
    }).then(() => {
      this.stopTreesLoaded = true
    })
  }

  setTransitiveNetwork (json) {
    return this.work({
      command: 'setTransitiveNetwork',
      message: {
        network: json
      }
    })
  }

  isReady () {
    return this.query && this.stopTreesLoaded
  }

  isLoaded () {
    return this.isReady() && this.originLoaded
  }

  getContour (cutoff = 60) {
    return this.work({
      command: 'getContour',
      message: {
        cutoff
      }
    })
  }

  /** Get a GeoJSON isochrone with the given cutoff (in minutes) */
  getIsochrone (cutoff = 60, interpolation = true) {
    return this.work({
      command: 'getIsochrone',
      message: {
        cutoff,
        interpolation
      }
    })
  }
}

export const createGrid = createGridFunc
