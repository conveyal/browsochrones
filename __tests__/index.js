/* global describe, expect, it */

import fs from 'fs'
import path from 'path'

import accessibilityForGrid from '../lib/accessibility-for-grid'
import generateDestinationData from '../lib/generate-destination-data'
import getIsochrone from '../lib/get-isochrone'
import {create as createSurface} from '../lib/surface'
import {create as createGrid} from '../lib/grid'
import {create as createOrigin} from '../lib/origin'
import {create as createStopTreeCache} from '../lib/stop-tree-cache'

const DATA = path.join(__dirname, '/data')
const DEST_POINT = {x: 158, y: 140}
const ORIGIN_POINT = {x: 125, y: 136}
const which = 'MEDIAN'

const jobsGrid = fs.readFileSync(`${DATA}/jobs.grid`).buffer
const originData = fs.readFileSync(`${DATA}/origin.dat`).buffer
const query = JSON.parse(fs.readFileSync(`${DATA}/query.json`, 'utf8'))
const stopTrees = fs.readFileSync(`${DATA}/stop_trees.dat`).buffer

const ctx = {}

describe('Browsochrones', () => {
  it('create stop tree cache', () => {
    ctx.stopTreeCache = createStopTreeCache({
      data: new Int32Array(stopTrees),
      size: query.width * query.height
    })
    const {index} = ctx.stopTreeCache
    expect(index.length).toMatchSnapshot()
    expect(index[0]).toMatchSnapshot()
    expect(index[index.length / 2]).toMatchSnapshot()
    expect(index[index.length - 1]).toMatchSnapshot()
  })

  it('create grid', () => {
    ctx.grid = createGrid(jobsGrid)
    const {data, ...snapshot} = ctx.grid
    expect(snapshot).toMatchSnapshot()
  })

  it('create origin', () => {
    ctx.origin = createOrigin({
      data: new Int32Array(originData),
      point: ORIGIN_POINT
    })
    const {data, index, ...snapshot} = ctx.origin
    expect(snapshot).toMatchSnapshot()
  })

  it('generate surface', () => {
    ctx.surface = createSurface({
      ...ctx,
      query,
      which
    })
    const {surface, ...snapshot} = ctx.surface
    expect(summarizeSnapshot(snapshot)).toMatchSnapshot()
  })

  it('generate destination data', () => {
    const snapshot = generateDestinationData({
      ...ctx,
      ...ctx.surface,
      from: ORIGIN_POINT,
      query,
      to: DEST_POINT,
      transitiveNetwork: query.transitiveData
    })
    expect(summarizeSnapshot(snapshot)).toMatchSnapshot()
  })

  it('get isochrone', () => {
    const isochrone = getIsochrone({
      ...query,
      surface: ctx.surface.surface
    })
    expect(summarizeSnapshot(isochrone)).toMatchSnapshot()
  })

  it('get accessibility for grid', () => {
    const access = accessibilityForGrid({
      grid: ctx.grid,
      query,
      surface: ctx.surface.surface
    })
    expect(access).toMatchSnapshot()
  })
})

function summarizeSnapshot (s, k) {
  if (Array.isArray(s) || ArrayBuffer.isView(s)) {
    return {
      [`first${k ? `-${k}` : ''}`]: summarizeSnapshot(s[0]),
      [`last${k ? `-${k}` : ''}`]: summarizeSnapshot(s[s.length - 1]),
      length: s.length
    }
  }
  if (typeof s === 'object') {
    const copy = {...s}
    Object.keys(copy).forEach((k) => {
      copy[k] = summarizeSnapshot(copy[k], k)
    })
    return copy
  }
  if (typeof s === 'number' && !Number.isInteger(s)) {
    return Number(s.toFixed(8)) // precision varies per environment (?)
  }
  return s
}
