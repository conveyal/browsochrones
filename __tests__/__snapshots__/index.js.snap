// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Browsochrones create grid 1`] = `
Object {
  "contains": [Function],
  "height": 237,
  "max": 12558,
  "min": 0,
  "north": 49616,
  "west": 33985,
  "width": 355,
  "zoom": 9,
}
`;

exports[`Browsochrones create origin 1`] = `
Object {
  "nMinutes": 120,
  "nStops": 3806,
  "radius": 20,
  "x": 125,
  "y": 136,
}
`;

exports[`Browsochrones create stop tree cache 1`] = `53320`;

exports[`Browsochrones create stop tree cache 2`] = `0`;

exports[`Browsochrones create stop tree cache 3`] = `1689608`;

exports[`Browsochrones create stop tree cache 4`] = `3469825`;

exports[`Browsochrones generate destination data 1`] = `
Object {
  "inVehicleTravelTime": 19,
  "paths": Object {
    "first-paths": Object {
      "first": 1995,
      "last": 0,
      "length": 2,
    },
    "last-paths": Object {
      "first": 1995,
      "last": 0,
      "length": 2,
    },
    "length": 120,
  },
  "times": Object {
    "first-times": 56,
    "last-times": 53,
    "length": 120,
  },
  "transitive": Object {
    "journeys": Object {
      "first-journeys": Object {
        "journey_id": 0,
        "journey_name": 0,
        "segments": Object {
          "first-segments": Object {
            "color": "rgb(165, 187, 0)",
            "from": Object {
              "place_id": "from",
              "type": "PLACE",
            },
            "to": Object {
              "stop_id": "1959",
              "type": "STOP",
            },
            "type": "WALK",
          },
          "last-segments": Object {
            "color": "rgb(165, 187, 0)",
            "from": Object {
              "stop_id": "3698",
              "type": "STOP",
            },
            "to": Object {
              "place_id": "to",
              "type": "PLACE",
            },
            "type": "WALK",
          },
          "length": 3,
        },
      },
      "last-journeys": Object {
        "journey_id": 1,
        "journey_name": 1,
        "segments": Object {
          "first-segments": Object {
            "color": "rgb(0, 187, 44)",
            "from": Object {
              "place_id": "from",
              "type": "PLACE",
            },
            "to": Object {
              "stop_id": "1959",
              "type": "STOP",
            },
            "type": "WALK",
          },
          "last-segments": Object {
            "color": "rgb(0, 187, 44)",
            "from": Object {
              "stop_id": "1995",
              "type": "STOP",
            },
            "to": Object {
              "place_id": "to",
              "type": "PLACE",
            },
            "type": "WALK",
          },
          "length": 3,
        },
      },
      "length": 2,
    },
    "patterns": Object {
      "first-patterns": Object {
        "pattern_id": "0",
        "pattern_name": "50",
        "route_id": "0",
        "stops": Object {
          "first-stops": Object {
            "geometry": "cvwqF~|}lOD{EnGH\`@kh@hEZpIJf@D\`GHbAkyAwGIy@voAE\`H@d@o@vt@k@CwLQ",
            "stop_id": "736",
          },
          "last-stops": Object {
            "geometry": Object {},
            "stop_id": "736",
          },
          "length": 17,
        },
      },
      "last-patterns": Object {
        "pattern_id": "382",
        "pattern_name": "3",
        "route_id": "35",
        "stops": Object {
          "first-stops": Object {
            "geometry": "irvqF\`kzlOt@@EjH",
            "stop_id": "582",
          },
          "last-stops": Object {
            "geometry": Object {},
            "stop_id": "2478",
          },
          "length": 25,
        },
      },
      "length": 383,
    },
    "places": Object {
      "first-places": Object {
        "place_id": "from",
        "place_lat": 39.76843641,
        "place_lon": -86.15203857,
        "place_name": undefined,
      },
      "last-places": Object {
        "place_id": "to",
        "place_lat": 39.75999141,
        "place_lon": -86.06140137,
        "place_name": undefined,
      },
      "length": 2,
    },
    "routes": Object {
      "first-routes": Object {
        "agency_id": "IndyGo",
        "route_color": "4269a4",
        "route_id": "63",
        "route_long_name": "Sunday North Loop",
        "route_short_name": "26SN",
        "route_type": 3,
      },
      "last-routes": Object {
        "agency_id": "IndyGo",
        "route_color": "4269a4",
        "route_id": "0",
        "route_long_name": "‚ÄùRed Line\\"",
        "route_short_name": "50",
        "route_type": 3,
      },
      "length": 64,
    },
    "stops": Object {
      "first-stops": Object {
        "stop_id": "0",
        "stop_lat": 39.8223119,
        "stop_lon": -85.990967,
        "stop_name": "Mitthoefer Rd & John Marshall Dr N",
      },
      "last-stops": Object {
        "stop_id": "3805",
        "stop_lat": 39.912093,
        "stop_lon": -86.186015,
        "stop_name": "86th St & Northbrook Shopping Center",
      },
      "length": 3806,
    },
  },
  "travelTime": 54,
  "waitTime": 7,
  "walkTime": 29,
}
`;

exports[`Browsochrones generate surface 1`] = `
Object {
  "inVehicleTravelTimes": Object {
    "first-inVehicleTravelTimes": 255,
    "last-inVehicleTravelTimes": 255,
    "length": 53320,
  },
  "spectrogramData": Object {
    "first-spectrogramData": Object {
      "first": 0,
      "last": 0,
      "length": 120,
    },
    "last-spectrogramData": Object {
      "first": 0,
      "last": 0,
      "length": 120,
    },
    "length": 120,
  },
  "waitTimes": Object {
    "first-waitTimes": 255,
    "last-waitTimes": 255,
    "length": 53320,
  },
  "walkTimes": Object {
    "first-walkTimes": 255,
    "last-walkTimes": 255,
    "length": 53320,
  },
}
`;

exports[`Browsochrones get accessibility for grid 1`] = `311967`;

exports[`Browsochrones get isochrone 1`] = `
Object {
  "geometry": Object {
    "coordinates": Object {
      "first-coordinates": Object {
        "first": Object {
          "first": Object {
            "first": -86.18270874,
            "last": 39.89920151,
            "length": 2,
          },
          "last": Object {
            "first": -86.18270874,
            "last": 39.89920151,
            "length": 2,
          },
          "length": 1041,
        },
        "last": Object {
          "first": Object {
            "first": -86.12182617,
            "last": 39.63319207,
            "length": 2,
          },
          "last": Object {
            "first": -86.12182617,
            "last": 39.63319207,
            "length": 2,
          },
          "length": 5,
        },
        "length": 34,
      },
      "last-coordinates": Object {
        "first": Object {
          "first": Object {
            "first": -86.25366211,
            "last": 39.71352536,
            "length": 2,
          },
          "last": Object {
            "first": -86.25366211,
            "last": 39.71352536,
            "length": 2,
          },
          "length": 5,
        },
        "last": Object {
          "first": Object {
            "first": -86.25366211,
            "last": 39.71352536,
            "length": 2,
          },
          "last": Object {
            "first": -86.25366211,
            "last": 39.71352536,
            "length": 2,
          },
          "length": 5,
        },
        "length": 1,
      },
      "length": 10,
    },
    "type": "MultiPolygon",
  },
  "type": "Feature",
}
`;
